### ISSUE TITLE
Questions regarding the architecture and dimensionality

### ISSUE THREAD

**Issue by lucgeo** [rwth-i6/returnn#163]
Hi,

I'm trying to analyze the architecture implemented in Returnn in a deep way: the type of layers, the input and the output of each one, the dimensionality, etc. I'm working on Librispeech recipe, I run the training process in Tensorboard step by step, I got the output of the "--task=analyze", I'm looking in the config file and I wanna put head-by-head all those things. However, some of them are still unclear:

* I'm sure that there are 6 Bi-LSTM layers (ids 0-5). I understand what is the input for the each one and their dimensionality. Which are the following ones and how many they are? I mean "ctc", "enc_ctx", "inv_fertility" - what is its role? Then it seems that there are several layers in the output block (I have listed them at the next point).

* In the output block, there are some layers (?) for which I don't know what is their input: energy, output_prob, readout_in, lstm_cell, s_transformed, target_embed.

* Related to the previous question, I don't understand the dimensionality of the weight matrices corresponding to the output_prob (500x10025). The 10025 value represents the number of acoustic states I guess, but 500? I can't link it with anything. Same problem for readout_in (3669x1000), lstm_cell (3669x4000), s_transformed (1000x1024) and target_embed (10025x621). Is the 621 the size of the context vector?

Many thanks!

**Comment by albertz** [comment]
I guess you refer to [this config](https://github.com/rwth-i6/returnn-experiments/blob/master/2018-asr-attention/librispeech/full-setup-attention/returnn.config), right?

The network as it is there in the config fully describes what is in the network: It is exactly what is defined there. I.e. you have `lstm%i_fw` for i in 0,...,5, you have one single `enc_ctx`, one single `ctc`, etc.

For understanding what they do, please also look at [the paper](https://www-i6.informatik.rwth-aachen.de/publications/download/1068/Zeyer--2018.pdf).

* `enc_ctx` corresponds to the attention energy calculation `e_i`, specifically the multiplication `W h_t` in there.
* `inv_fertility` is used for the weight feedback `beta`, specifically the `sigma(...)`.
* `energy` is exactly the final result of `e_i` in the paper. The input of this is the result from `tanh(...)`. The layer itself is performing the linear transformation, i.e. the multiplication with `v`.
* `output_prob` is `p(y_i|...)` from the paper, i.e. the probability distribution over all the output classes. The layer basically does the final `linear` and `softmax` from the paper. The input to this is `readout`, which corresponds to the result of `maxout` in the paper. `readout` has dimension 500.
* `readout_in` gets `s` (`s_i` in paper), `prev:target_embed` (`y_{i-1}` in paper), `att` (`c_i` in paper) as input. These are concatenated, i.e. the input dimension is 1000 + 621 + 2048 = 3669.
* `lstm_cell` is `s_i` in the paper. It get's in `prev:target_embed` and `prev:att`, i.e. input dimension is 621 + 2048. But then it also has the recurrent part, i.e. itself, which is another 1000 dimensions.
* `s_transformed` is basically `W s_i` in the paper, which is used again for `enc_ctx`. 
* `target_embed`: Basically `y_i`. But `y_i` are just indices, so you need an embedding of them. This is it. In the config, you see `"target_embed": {..., "n_out": 621}`, i.e. that defines that it has 621 dimensions. And there are 10025 possible labels.

You have `debug_print_layer_output_template` in that config. This means, when you run RETURNN, it will print you detailed information about the shape, dimensions, etc of each layer. This is probably also helpful in understanding. Please check that.

**Comment by albertz** [comment]
I'm closing now as this is not really an issue with RETURNN. Feel free to ask further questions, though.

**Comment by lucgeo** [comment]
Hi,

Thank you so much for your explanations! I still have some other questions:

* if there are 10025 output labels, why the value of 10026 is it involved in the ctc layer? Is it somehow 10025 labels + blank character, which is necessary in ctc?

* you said that readout has dimension 500, but I know the weight matrix involved in the readout_in layer has dimension [3669,100]. How does the 1000 value become 500? Is the maxout doing that?

* Which was the criteria to set the size of the embed vector as 621?

**Comment by albertz** [comment]
> * if there are 10025 output labels, why the value of 10026 is it involved in the ctc layer? Is it somehow 10025 labels + blank character, which is necessary in ctc?

Yes exactly.

> * you said that readout has dimension 500, but I know the weight matrix involved in the readout_in layer has dimension [3669,100]. How does the 1000 value become 500? Is the maxout doing that?

Check the output (via `debug_print_layer_output_template`). You will see all the shape/dim information. Esp, you will see exactly that the maxout is doing that (layer `readout`, `"class": "reduce_out", "mode": "max", "num_pieces": 2`).

> * Which was the criteria to set the size of the embed vector as 621?

I never changed it. This was our machine translation baseline.

**Comment by lucgeo** [comment]
Hi,

1. I have doubts regarding the sequence length. Which one of 176, 55 and 75 is the sequence length? There is not a fixed value?
"<LibriSpeechCorpus 'train' epoch=1>, epoch 1. Old mean seq len (transcription) is 176.233378, new is 55.463026, requested max is 75.000000. Old num seqs is 14063, new num seqs is 1758"

2. The values below I think they represent the number of sequences per batch. Am I right?
"Train data: LibriSpeechCorpus, sequences: 1758, frames: unknown
Dev data: LibriSpeechCorpus, sequences: 3000, frames: unknown"

3. Are there some language dependent parameters? For instance, I mean that the number of phonemes for a specific language is related with the number of the output labels.

Thank you!

**Comment by albertz** [comment]
> 1. I have doubts regarding the sequence length. Which one of 176, 55 and 75 is the sequence length? There is not a fixed value?
>    "<LibriSpeechCorpus 'train' epoch=1>, epoch 1. Old mean seq len (transcription) is 176.233378, new is 55.463026, requested max is 75.000000. Old num seqs is 14063, new num seqs is 1758"

They are all sequence lengths. We apply some filtering by sequence length in the early epochs (via option `epoch_wise_filter`).
See the code for full details (`LibriSpeechCorpus.init_seq_order`).

> 2. The values below I think they represent the number of sequences per batch. Am I right?
>    "Train data: LibriSpeechCorpus, sequences: 1758, frames: unknown
>    Dev data: LibriSpeechCorpus, sequences: 3000, frames: unknown"

No, they represent the number of sequences for the first epoch.
(Train data is maybe a bit misleading there, because this is only the first epoch, not the full train data. And the first epoch specifically is even further filtered a lot by sequence length.)

> 3. Are there some language dependent parameters? For instance, I mean that the number of phonemes for a specific language is related with the number of the output labels.

For what exactly? You see all parameters which we used for LibriSpeech in the config. There is not much related to the language. Maybe the size of the BPE alphabet.

**Comment by lucgeo** [comment]
For training and decoding using this framework for another language than English. Yeah, the BPE alphabet should be different. I was wondering if there are other language dependent parameters.
Thank you!

**Comment by albertz** [comment]
There is nothing directly dependent on the language. But I can imagine that some of the hyper params probably need some tuning again, like the embedding sizes, etc.

**Comment by lucgeo** [comment]
Hi,

Hi,
From this config file [1], the weight feedback (Beta) is a 1024 dimensional vector. But according to the formula from here [2], it is obtained as a multiplication of sigma(inv_fertility) and sum(attention weights). From my understading, both those factors are scalars. Is not possible to multiply two scalars, obtaining a vector. Where am I wrong? Which of them is a 1024 dimensional vector?

Many thanks!

[1]: https://github.com/rwth-i6/returnn-experiments/blob/master/2018-asr-attention/librispeech/full-setup-attention/returnn.config
[2]: https://www-i6.informatik.rwth-aachen.de/publications/download/1068/Zeyer--2018.pdf

**Comment by albertz** [comment]
Again, you should look at the output of RETURNN (with `debug_print_layer_output_template`), which should clarify most of these questions.

There is
`"weight_feedback": {"class": "linear", "activation": None, "with_bias": False, "from": ["prev:accum_att_weights"], "n_out": EncKeyTotalDim}`,
and `prev:accum_att_weights` here is a scalar as you mention (or rather it is actually a vector of dimension 1; again, check the output of RETURNN which will print all that information)
(or rather, actually that is for all encoder time frames, i.e. you have shape `(enc_time, 1)`, which is exactly what RETURNN tells you in the output).
This is a `linear` layer, i.e. it will do a matrix multiplication (matrix of shape `(input_dim, output_dim)`, where input_dim is 1 here, and output_dim is the `n_out` setting).
I.e. the output of this layer is a matrix of shape `(enc_time, EncKeyTotalDim)`. (This is also what RETURNN will tell you in the output.)

This is also the same as in the paper. You have the scalar (or 1-dim vector) `beta_{i,t}`, and that one is then multiplied by some matrix (for `e_{i,t}`).

**Comment by lucgeo** [comment]
Hi,
I know, I'm using 'debug_print_layer_output_template', but this toolkit is new for me and even some things seems obvious, I can't catch them right away or I want to be sure if I'm right.

Coming back to the question about the sequence length, which is the approximate number of frames per chunk (sequence) that could be considered? How many inputs are processed at a time? I need it to compute the number of operations. 

Thank you!

**Comment by albertz** [comment]
You can configure that via batch size (total number of frames in a batch), max seqs (in a batch), and max seq length. You see the final result in the training output on stdout.

**Comment by lucgeo** [comment]
Hi,

In the paper it's mentioned that " After every layer in the encoder, we optionally do max-pooling in the time dimension to reduce the encoder length". 

We have this row in the config file:
"lstm0_pool": {"class": "pool", "mode": "max", "padding": "same", "pool_size": (2,), "from": ["lstm0_fw", "lstm0_bw"], "trainable": False}

1. Does it have that pool size parameter a similar role to a stride? I mean, applying a pool_size(2,) over a 200 frames sequence, it will result in 100 frames?
In this case, the time dimension is reduced only in the first 3 encoder layers for [this] implementation, the the following ones are retaining the same sequence length. Am I right?

2. What's the actual scope of the 'layer wise filter'? Looking to the source code in MetaDataset.py, it looks for me like a sequence length normalization...

Thank you!

[this]: https://github.com/rwth-i6/returnn-experiments/blob/master/2018-asr-attention/librispeech/full-setup-attention/returnn.config

**Comment by albertz** [comment]
This uses the `PoolLayer` (see implementation for details). The stride for this layer is by default the same as the pool size.

Yes, in that config, you have 3 times a pool/stride size of 2, i.e. a time reduction factor 8 in total. The other pool layers with pool/stride size of 1 do nothing.

I guess you mean `epoch_wise_filter`, not 'layer wise filter'. Not sure what you mean by scope. Yes, as said, this filters out some sequences from the training data such that the mean sequence length is below some threshold.

**Comment by lucgeo** [comment]
Yes, I was tired and I expressed myself wrong. I mean 'epoch_wise_filter' and I was asking about its purpose, not scope.
These are the answers I wanted. Thank you!

**Comment by lucgeo** [comment]
Hi,

I want to know which is the sequence length processed at a time in this [experiment]. More exactly, I'm interested in the number of audio frames served as the network input. 

Thank you!


[experiment]: https://github.com/rwth-i6/returnn-experiments/blob/master/2018-asr-attention/librispeech/full-setup-attention/returnn.config

**Comment by albertz** [comment]
In that experiment, the encoder is a bidirectional LSTM. I.e. every single frame of the encoder output has seen the whole input.
